rfc-134-govuk-wide-session-cookie-and-login.md

```diff
@@ -0,0 +1,337 @@
+# A GOV.UK-wide session cookie & login
+
+## Summary
+
+We propose introducing a new cookie, `govuk_session`, which will be an
+optional cookie, required only for users who wish to use personalised
+parts of GOV.UK (currently just the Transition Checker).
+
+Similarly to how our A/B tests work we will manage this cookie at the
+Fastly layer, in Varnish Configuration Language (VCL), and use custom
+request headers to pass the cookie value to our apps. We will also use
+custom response headers to set a new cookie value.
+
+We will create a new app to handle the login and logout process, and
+to manage the OAuth tokens used to update a user's data.
+
+## Problem
+
+The GOV.UK Account team have launched an experiment on the Transition
+Checker, allowing users to sign up to save their
+results. [finder-frontend sets a session cookie][ff-session]—an
+encrypted cookie—containing the user ID and OAuth tokens used to
+update the data we hold on them.
+
+This works fine for one app, but there are problems with this approach
+when we try to scale to more than one app.
+
+[ff-session]: https://github.com/alphagov/finder-frontend/blob/5b6101e8eb027028a102ab96bcef6ff752b9c558/app/controllers/sessions_controller.rb#L23-L28
+
+### We need to pass cookies to our apps
+
+Our Nginx configuration blocks most cookies, [which we disabled for
+the Transition Checker][tc-cookie-pr]. If multiple apps need to
+consume the session cookie, then this blocking will be disabled for
+ever-increasing chunks of GOV.UK.
+
+[tc-cookie-pr]: https://github.com/alphagov/govuk-puppet/pull/10788
+
+### Apps need to share the same session cookie
+
+It's no good if the user has to log into each part of GOV.UK
+individually.  For example, say we personalise taxon pages: a user
+shouldn't have to log into finder-frontend (for the Transition
+Checker) and collections (for the taxon pages) separately.
+
+There needs to be one session shared across them all.  If we use Rails
+session cookies for this, we need to make sure all apps use the same
+encryption key.
+
+### Which app handles logging in and out?
+
+If there is one session cookie used for all of GOV.UK, which app sets
+that?
+
+Somewhere there needs to be a login controller and a logout controller
+which manipulate the cookie. These controllers will redirect the user
+to the GOV.UK Account system to do the actual authentication, but we
+need something on www.gov.uk itself to set the cookie.
+
+### Non-personalised parts of GOV.UK need to manipulate the session cookie too
+
+It's unlikely that personalisation will roll out to all of GOV.UK in
+one go.  There will be islands of personalised content.  Currently
+there is the Transition Checker.  Maybe next will be some guidance
+pages, or something else.  We want to keep the user's session alive
+while they are browsing the non-personalised parts of GOV.UK,
```

Comment:
Should we think about services here? as users don't realise that service domains are different. Although I can't think of a simple fix.

Reply:
I think things off www.gov.uk are out of the scope of this RFC, as they fall under the cross-gov stuff (which is looking likely to sit with Digital Identity?)

Reply:
Out of interest, would Digital Identity cover the brokering aspect to an account? For example, a logged-in user using a service which needs the user's stored details brokered through the account. If those details can be stored without needing to verify the user's identity, does that fall under Digital Identity's remit?

Reply:
Currently it's looking like Digital Identity will be doing the cross-gov attribute exchange stuff, yes.  But the details are still being worked out.